File: c/algo/kadane.c
#include <stdio.h>
#include <limits.h>

/*PROTOTYPE*/
int kadane(int *, int);

/*DRIVE CODE*/
int main(){
    int arr[] = {-2, -3, 4, -1, -2, 1, 5, -3};
    int size = sizeof(arr) / sizeof(arr[0]);
    printf("Suma maxima: %d\n", kadane(arr, size));

    return 0;
}

/*FUNCTION*/
int kadane(int *arr, int size){
    int max_so_far = INT_MIN;
    int max_ending_here = 0;

    for(int i = 0; i < size; i++){
        max_ending_here += arr[i];
        if(max_so_far < max_ending_here) max_so_far = max_ending_here;
        if(max_ending_here < 0) max_ending_here = 0;
    }
    return max_so_far;
}

File: python/kadane.py
def kadane(arr):
    size = len(arr)
    max_so_far = float('-inf')
    max_ending_here = 0

    for i in range(size):
        max_ending_here += arr[i]
        if(max_so_far < max_ending_here): max_so_far = max_ending_here
        if(max_ending_here < 0): max_ending_here = 0
    
    return max_so_far


arr = []
while(True):
    value = input("Ingrese un numero o escriba 'Fin' para terminar: ")
    if value.lower() == 'fin':
        break
    try:
        num = int(value)
        arr.append(num)
    except ValueError:
        print("Ingrese un numero valido.")


print("Suma maxima: ", kadane(arr))